FastSurferCNN/utils/metrics.py

```diff
@@ -103,175 +107,317 @@ def precision_recall(
+        self.union = torch.zeros(self.n_classes, self.n_classes)
+        self.intersection = torch.zeros(self.n_classes, self.n_classes)
 
     def _check_output_type(self, output):
         """
-        Check the output type.
+        Check the type of the output and raise an error if it doesn't match expectations.
+
+        Parameters:
+        -----------
+        output : tuple
+            The output to be checked, expected to be a tuple.
         """
         if not (isinstance(output, tuple)):
             raise TypeError(
-                f"Output should a tuple consist of of torch.Tensors, but given {type(output)}"
+                f"Output should be a tuple consisting of torch.Tensors, but given {type(output)}"
             )
 
-    def _update_union_intersection_matrix(
-        self, batch_output: torch.Tensor, labels_batch: torch.Tensor
-    ):
+    def _update_union_intersection(self, batch_output, labels_batch):
         """
-        Update the union intersection matrix.
+        Update the union and intersection matrices based on batch predictions and labels.
 
-        Parameters
-        ----------
+        Parameters:
+        -----------
         batch_output : torch.Tensor
-            Output tensor.
-        labels_batch : torch.Tensor
-            Label batch.
-        """
-        for i in range(self.n_classes):
-            gt = (labels_batch == i).float()
-            for j in range(self.n_classes):
-                pred = (batch_output == j).float()
-                self.intersection[i, j] += torch.sum(torch.mul(gt, pred))
-                self.union[i, j] += torch.sum(gt) + torch.sum(pred)
-
-    def _update_union_intersection(
-        self, batch_output: torch.Tensor, labels_batch: torch.Tensor
-    ):
-        """
-        Update the union intersection.
+            Batch predictions from the model.
 
-        Parameters
-        ----------
-        batch_output : torch.Tensor
-            Batch output (prediction, labels).
-        labels_batch : torch.Tensor
+        labels_batch : np.ndarray or torch.Tensor
+            Batch labels from the dataset.
         """
-        for i in range(self.n_classes):
-            gt = (labels_batch == i).float()
-            pred = (batch_output == i).float()
-            self.intersection[i, i] += torch.sum(torch.mul(gt, pred))
-            self.union[i, i] += torch.sum(gt) + torch.sum(pred)
-
-    def update(self, output: tuple[Any, Any], cnf_mat: bool):
+        # self.union.to(batch_output.device)
+        # self.intersection.to(batch_output.device)
+        for i, c1 in enumerate(self.class_ids):
+            gt = (labels_batch == c1).float()
+            for j, c2 in enumerate(self.class_ids):
+                pred = (batch_output == c2).float()
+                self.intersection[i, j] = self.intersection[i, j] + torch.sum(
+                    torch.mul(gt, pred)
+                )
+                self.union[i, j] = self.union[i, j] + torch.sum(gt) + torch.sum(pred)
```

Comment:
Also, it seems like the original implementation had a condition here that would only compute the full confusion matrix if explicitly required. Most times, the diagonal of the confusion matrix is enough (see previous). And always computing the full confusion matrix is EXTREMELY computationally expensive. (Prohibitively so). So we should only compute the full confusion matrix if it is explicitly requested.